File: generators/paint_calculator.py
from data.publisher import CALCULATOR_FAMILIES
from generators.publisher_pages import render_calculator_page


def build_paint_calculator_page() -> str:
    family = next(item for item in CALCULATOR_FAMILIES if item["slug"] == "paint-calculator")
    form_html = """<div class="toggle-row" role="tablist" aria-label="Paint mode"><button class="mode-toggle is-active" data-mode="walls" type="button">Walls</button><button class="mode-toggle" data-mode="ceiling" type="button">Ceiling</button><button class="mode-toggle" data-mode="single" type="button">Single surface</button></div><div class="toggle-row units-row" role="tablist" aria-label="Units"><button class="unit-toggle is-active" data-unit="metric" type="button">Metric</button><button class="unit-toggle" data-unit="imperial" type="button">Imperial</button></div><form id="paint-form" class="calculator-form"><div class="field-grid"><label><span>Length</span><input id="length" type="number" min="0" step="0.01" value="4"></label><label><span>Width</span><input id="width" type="number" min="0" step="0.01" value="3"></label><label><span>Height</span><input id="height" type="number" min="0" step="0.01" value="2.4"></label><label><span>Coats</span><input id="coats" type="number" min="1" step="1" value="2"></label><label><span>Coverage per litre (m²)</span><input id="coverage" type="number" min="0" step="0.1" value="12"></label><label><span>Waste allowance (%)</span><input id="waste" type="number" min="0" step="1" value="10"></label><label><span>Price per litre</span><input id="price-per-litre" type="number" min="0" step="0.01" value="8.50"></label></div><button class="btn btn-primary btn-block" type="submit">Calculate paint</button></form>"""
    result_html = """<div class="result-kicker">Estimated result</div><h2 class="result-main">Enter your measurements</h2><p class="result-sub">You will see the painted area, litres, tin suggestion and rough material cost here.</p><div class="currency-pills" role="tablist" aria-label="Currency"><button class="currency-pill is-active" data-currency="GBP" type="button">GBP £</button><button class="currency-pill" data-currency="USD" type="button">USD $</button><button class="currency-pill" data-currency="EUR" type="button">EUR €</button></div><div class="result-breakdown" id="result-breakdown"></div>"""
    return render_calculator_page(
        slug=family["slug"],
        title=family["name"],
        description=family["meta_description"],
        intro=family["intro"],
        form_html=form_html,
        result_html=result_html,
        script_name="paint-calculator.js",
    )

File: assets/js/paint-calculator.js
(function () {
  const form = document.getElementById("paint-form");
  if (!form) return;

  const modeButtons = Array.from(document.querySelectorAll(".mode-toggle"));
  const unitButtons = Array.from(document.querySelectorAll(".unit-toggle"));
  const currencyButtons = Array.from(document.querySelectorAll(".currency-pill"));
  const resultMain = document.querySelector(".result-main");
  const resultSub = document.querySelector(".result-sub");
  const resultBreakdown = document.getElementById("result-breakdown");

  const currencyMap = {
    GBP: { symbol: "GBP £", rate: 1 },
    USD: { symbol: "USD $", rate: 1.27 },
    EUR: { symbol: "EUR €", rate: 1.17 },
  };

  let mode = "walls";
  let unit = "metric";
  let currency = "GBP";

  function getNumber(id) {
    const field = document.getElementById(id);
    const value = field ? parseFloat(field.value) : 0;
    return Number.isFinite(value) ? value : 0;
  }

  function formatMoney(value) {
    const info = currencyMap[currency] || currencyMap.GBP;
    return `${info.symbol} ${(value * info.rate).toFixed(2)}`;
  }

  function setActive(buttons, key, value) {
    buttons.forEach(function (button) {
      button.classList.toggle("is-active", button.dataset[key] === value);
    });
  }

  function toMetricLength(value) {
    return unit === "metric" ? value : value * 0.3048;
  }

  function currentArea(length, width, height) {
    if (mode === "ceiling") {
      return length * width;
    }
    if (mode === "single") {
      return length * height;
    }
    return (2 * (length + width)) * height;
  }

  function modeLabel() {
    if (mode === "ceiling") return "ceiling area";
    if (mode === "single") return "single surface";
    return "wall area";
  }

  function calculate(event) {
    if (event) event.preventDefault();

    const length = toMetricLength(getNumber("length"));
    const width = toMetricLength(getNumber("width"));
    const height = toMetricLength(getNumber("height"));
    const coats = Math.max(getNumber("coats"), 1);
    const coverage = Math.max(getNumber("coverage"), 0.1);
    const waste = Math.max(getNumber("waste"), 0);
    const pricePerLitre = Math.max(getNumber("price-per-litre"), 0);

    const baseArea = currentArea(length, width, height);
    const adjustedArea = baseArea * coats * (1 + waste / 100);
    const litres = adjustedArea / coverage;
    const estimatedCost = litres * pricePerLitre;

    if (!(litres > 0)) {
      resultMain.textContent = "Enter your measurements";
      resultSub.textContent = "You will see the painted area, litres, tin suggestion and rough material cost here.";
      resultBreakdown.innerHTML = "";
      return;
    }

    const tenLitreTins = Math.floor(litres / 10);
    const remainderAfterTen = litres - (tenLitreTins * 10);
    const fiveLitreTins = Math.floor(remainderAfterTen / 5);
    const topUpLitres = Math.max(remainderAfterTen - (fiveLitreTins * 5), 0);

    resultMain.textContent = `${litres.toFixed(2)} litres of paint`;
    resultSub.textContent = `That covers about ${adjustedArea.toFixed(2)} m² across ${coats} coat(s), with a rough material cost of ${formatMoney(estimatedCost)}.`;
    resultBreakdown.innerHTML = `
      <div class="break-row"><span>Estimated ${modeLabel()}</span><strong>${baseArea.toFixed(2)} m²</strong></div>
      <div class="break-row"><span>Area incl. coats and waste</span><strong>${adjustedArea.toFixed(2)} m²</strong></div>
      <div class="break-row"><span>Coverage per litre</span><strong>${coverage.toFixed(2)} m²/L</strong></div>
      <div class="break-row"><span>Paint needed</span><strong>${litres.toFixed(2)} litres</strong></div>
      <div class="break-row"><span>Price per litre</span><strong>${formatMoney(pricePerLitre)}</strong></div>
      <div class="break-row"><span>Estimated material cost</span><strong>${formatMoney(estimatedCost)}</strong></div>
      <div class="calc-note">Suggested buying mix: ${tenLitreTins ? `${tenLitreTins} × 10L` : "0 × 10L"}${fiveLitreTins ? `, ${fiveLitreTins} × 5L` : ""}${topUpLitres > 0.05 ? `, plus a small top-up tin` : ""}.</div>
    `;
  }

  modeButtons.forEach(function (button) {
    button.addEventListener("click", function () {
      mode = button.dataset.mode || "walls";
      setActive(modeButtons, "mode", mode);
      calculate();
    });
  });

  unitButtons.forEach(function (button) {
    button.addEventListener("click", function () {
      unit = button.dataset.unit || "metric";
      setActive(unitButtons, "unit", unit);
      calculate();
    });
  });

  currencyButtons.forEach(function (button) {
    button.addEventListener("click", function () {
      currency = button.dataset.currency || "GBP";
      setActive(currencyButtons, "currency", currency);
      calculate();
    });
  });

  form.addEventListener("submit", calculate);
  form.querySelectorAll("input").forEach(function (input) {
    input.addEventListener("input", calculate);
  });

  calculate();
})();
